### ISSUE TITLE
[MENFORCER-393] Upgrading to 3.0.0 causes `Could not build dependency tree` with repositories some unknown protocol

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-enforcer#512]
**[johnny willer gasperi goncalves](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=johnnywiller)** opened **[MENFORCER-393](https://issues.apache.org/jira/browse/MENFORCER-393?redirect=false)** and commented

After upgrading to 3.0.0, it's not possible to validate the POM anymore, an error like 

```java
Could not build dependency tree Could not collect dependencies: {jarname}
```

happens.

 

I'm attaching the `mvn validate -X` dump (i have omitted some jars from the output) 

---

**Affects:** 3.0.0

**Attachments:**
- [enforcer_output](https://issues.apache.org/jira/secure/attachment/13031649/enforcer_output) (_123.83 kB_)

**Remote Links:**
- [GitHub Pull Request #271
](https://github.com/apache/maven-enforcer/pull/271)

3 votes, 10 watchers

**Comment by jira-importer** [comment]
**[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz)** commented

Can you verify if the wagon-s3 extension setup is correct?

 

`[DEBUG] Failure to find com.lastmile.finance:funding-instrument-ws:10.0.22-SNAPSHOT/maven-metadata.xml in https://nexus.atm.osp.tech/repository/maven-snapshot-group/ was cached in the local repository, resolution will not be reattempted until the update interval of atm-osp-payments-snapshot has elapsed or updates are forced`
`[DEBUG] No transport java.util.NoSuchElementException`
` role: org.apache.maven.wagon.Wagon`
` roleHint: s3`
`org.codehaus.plexus.component.repository.exception.ComponentLookupException: java.util.NoSuchElementException`
` role: org.apache.maven.wagon.Wagon`
` roleHint: s3`

**Comment by jira-importer** [comment]
**[johnny willer gasperi goncalves](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=johnnywiller)** commented

Hi Sylwester, thank you very much to comment on this.

We are not using wagon, but I see Maven Enforcer uses it, i didn't see on the usage page a need for configure wagon, is that necessary? can you give more info on how to setup that?

 

When using Maven enforcer v1.4.1 this was not needed

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

[johnny willer gasperi goncalves](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=johnnywiller) any reproducer?

by the way I can reproduce the same issue with this dependency:

\<dependency>
\<groupId>org.jruby\</groupId>
\<artifactId>jruby-stdlib\</artifactId>
\<version>9.2.13.0\</version>
\</dependency>

 

which have 

 
\<repositories>
\<repository>
\<id>mavengems\</id>
\<url>mavengem:https://rubygems.org\</url>
\</repository>
\</repositories>

**Comment by jira-importer** [comment]
**[johnny willer gasperi goncalves](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=johnnywiller)** commented

Hi [Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy), sorry I don't have any reproducer, the library in case is private one. I'll do some tests, If I find one i share here. Thanks for looking at it

**Comment by jira-importer** [comment]
**[Quentin Caillard](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dichotomia)** commented

I have this message when upgrading to 3.0.0 from 3.0.0-M3

[WARNING] Rule 5: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Could not build dependency tree Could not collect dependencies: com.mac:mac-bc360-custom:jar:9.3.2-SNAPSHOT ()

**Comment by jira-importer** [comment]
**[Marcos Passos](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=marcospassos)** commented

Exactly the same issue here.

**Comment by jira-importer** [comment]
**[Tilman Hausherr](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tilman)** commented

Same problem in the Apache Tika project ( https://github.com/apache/tika/ ). I was trying to change version from 3.0.0-M3 to 3.0.0 in the parent pom.xml and one of the subprojects fails.

Old output

```
--- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-maven-version) @ tika-parser-scientific-module ---

--- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-java-version) @ tika-parser-scientific-module ---

--- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-output-timestamp-property) @ tika-parser-scientific-module ---

--- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-maven) @ tika-parser-scientific-module ---

--- maven-enforcer-plugin:3.0.0-M3:enforce (ban-bad-log4j-versions) @ tika-parser-scientific-module ---

```

New output

```
--- maven-enforcer-plugin:3.0.0:enforce (enforce-maven-version) @ tika-parser-scientific-module ---

--- maven-enforcer-plugin:3.0.0:enforce (enforce-java-version) @ tika-parser-scientific-module ---

--- maven-enforcer-plugin:3.0.0:enforce (enforce-output-timestamp-property) @ tika-parser-scientific-module ---

--- maven-enforcer-plugin:3.0.0:enforce (enforce-maven) @ tika-parser-scientific-module ---
Rule 0: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
Could not build dependency tree Could not collect dependencies: org.apache.tika:tika-parser-scientific-module:jar:2.2.2-SNAPSHOT ()
```

**Comment by jira-importer** [comment]
**[Tilman Hausherr](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tilman)** commented

Maybe related: before the above comment, I had to change two entries in another subproject due to convergence problems, like this because osgi 6 was used elsewhere, and log4j has a weak dependency on an earlier version of osgi:

```
<dependency>
  <groupId>org.apache.logging.log4j</groupId>
  <artifactId>log4j-core</artifactId>
  <version>${log4j2.version}</version>
  <scope>test</scope>
  <exclusions>
      <!-- depends on org.osgi.core 4.3.1 -->
      <exclusion>
          <groupId>org.osgi</groupId>
          <artifactId>org.osgi.core</artifactId>
      </exclusion>
  </exclusions>
</dependency>
```

**Comment by jira-importer** [comment]
**[Tilman Hausherr](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tilman)** commented

Still happens with 3.1.0.

**Comment by jira-importer** [comment]
**[Falko Modler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=famod)** commented

Same here with enforcer 3.1.0 and Maven 3.8.6:

```
[ERROR] Rule 0: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
Could not build dependency tree Could not collect dependencies: io.quarkus:quarkus-extension-processor:jar:999-SNAPSHOT
```

`-X` reveals:

```
...
Caused by: org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact rubygems:rubygems-update:pom:3.2.29 from/to mavengems (mavengem:https://rubygems.org): Cannot access mavengem:https://rubygems.org with type default using the available connector factories: BasicRepositoryConnectorFactory
...
Caused by: org.eclipse.aether.transfer.NoTransporterException: Unsupported transport protocol mavengem:https
...
Caused by: org.codehaus.plexus.component.repository.exception.ComponentLookupException: java.util.NoSuchElementException
      role: org.apache.maven.wagon.Wagon
  roleHint: mavengem
Caused by: java.util.NoSuchElementException
    at org.eclipse.sisu.plexus.RealmFilteredBeans$FilteredItr.next (RealmFilteredBeans.java:118)
    at org.eclipse.sisu.plexus.RealmFilteredBeans$FilteredItr.next (RealmFilteredBeans.java:1)
    at org.eclipse.sisu.plexus.DefaultPlexusBeans$Itr.next (DefaultPlexusBeans.java:76)
    at org.eclipse.sisu.plexus.DefaultPlexusBeans$Itr.next (DefaultPlexusBeans.java:1)
    at org.codehaus.plexus.DefaultPlexusContainer.lookup (DefaultPlexusContainer.java:263)
    at org.codehaus.plexus.DefaultPlexusContainer.lookup (DefaultPlexusContainer.java:255)
    at org.eclipse.aether.internal.transport.wagon.PlexusWagonProvider.lookup (PlexusWagonProvider.java:56)
    ...
```

Ruby comes in via: https://github.com/quarkusio/quarkus/blob/main/core/processor/pom.xml#L42-L46

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

I added `org.jruby:jruby-stdlib` to IT - and looks like test pass.

Please retest with 3.3.0 version

**Comment by jira-importer** [comment]
**[Tilman Hausherr](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tilman)** commented

The problem I mentioned in January 2022 no longer happens with 3.3.0.

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

fixed in 3.3.0 - now I added IT for regressions, so reports in 3.3.1
